3p/integration.js

```diff
@@ -113,6 +120,51 @@ function nonSensitiveDataPostMessage(type, opt_object) {
   var object = opt_object || {};
   object.type = type;
   object.sentinel = 'amp-3p';
-  // Use of * is OK. We are not worried who gets this message.
-  window.parent./*OK*/postMessage(object, '*');
+  window.parent./*OK*/postMessage(object,
+      window.context.location.origin);
+}
+
+/**
+ * Registers a callback for intersections of this iframe with the current
+ * viewport.
+ * The passed in array has entries that aim to be compatible with
+ * the IntersectionObserver spec callback.
+ * http://rawgit.com/slightlyoff/IntersectionObserver/master/index.html#callbackdef-intersectionobservercallback
+ * @param {function(!Array<IntersectionObserverEntry>)} observerCallback
+ */
+function observeIntersection(observerCallback) {
+  // Send request to received records.
+  nonSensitiveDataPostMessage('send-intersections');
+  window.addEventListener('message', function(event) {
+    if (event.source != window.parent ||
+        event.origin != window.context.location.origin ||
+        !event.data ||
+        event.data.sentinel != 'amp-3p' ||
+        event.data.type != 'intersection') {
+      return;
+    }
+    observerCallback(event.data.changes);
+  });
+}
+
+/**
+ * Throws if the current frame's parent origin is not equal to
+ * the claimed origin.
+ * For browsers that don't support ancestorOrigins it adds
+ * `originValidated = false` to the location object.
+ * @param {!Window} window
+ * @param {!Location} parentLocation
+ * @visibleForTesting
+ */
+export function validateParentOrigin(window, parentLocation) {
+  var ancestors = window.location.ancestorOrigins;
+  if (ancestors == null || !ancestors.length) {
+    parentLocation.originValidated = false;
```

Comment:
So, this is off for the browsers that don't have ancestorOrigins, right? If so, it'd probably be cleaner to check of `!ancestors` given that `undefined` is the real value for us here.